src/OpenTelemetry/Metrics/ThreadStaticStorage.cs

```diff
@@ -77,52 +77,79 @@ internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> ta
-                {
-                    i++;
-                }
-            }
-
-            actualLength = i;
-
-            if (actualLength == 0)
+            // We do not know ahead the actual length, so start with max possible length.
+            var maxLength = Math.Min(tagKeysInteresting.Count, tagLength);
+            if (maxLength == 0)
             {
                 tagKeys = null;
                 tagValues = null;
             }
-            else if (actualLength <= MaxTagCacheSize)
+            else if (maxLength <= MaxTagCacheSize)
             {
-                tagKeys = this.primaryTagStorage[actualLength - 1].TagKeys;
-                tagValues = this.primaryTagStorage[actualLength - 1].TagValues;
+                tagKeys = this.primaryTagStorage[maxLength - 1].TagKeys;
+                tagValues = this.primaryTagStorage[maxLength - 1].TagValues;
             }
             else
             {
-                tagKeys = new string[actualLength];
-                tagValues = new object[actualLength];
+                tagKeys = new string[maxLength];
+                tagValues = new object[maxLength];
             }
 
-            // Iterate again (!) to assign the actual value.
-            // TODO: The dual iteration over tags might be
-            // avoidable if we change the tagKey and tagObject
-            // to be a different type (eg: List).
-            // It might lead to some wasted memory.
-            // Also, it requires changes to the Dictionary
-            // used for lookup.
-            // The TODO here is to make that change
-            // separately, after benchmarking.
-            i = 0;
+            actualLength = 0;
             for (var n = 0; n < tagLength; n++)
             {
+                // Copy only interesting tags, and keep count.
                 var tag = tags[n];
                 if (tagKeysInteresting.Contains(tag.Key))
                 {
-                    tagKeys[i] = tag.Key;
-                    tagValues[i] = tag.Value;
-                    i++;
+                    tagKeys[actualLength] = tag.Key;
+                    tagValues[actualLength] = tag.Value;
+                    actualLength++;
+                }
+            }
+
+            // If the actual length was equal to max, great!
+            // else, we need to pick the array of the actual length,
+            // and copy tags into it.
+            // This optimizes the common scenario:
+            // User is interested only in TagA and TagB
+            // and incoming measurement has TagA and TagB and many more.
+            // In this case, the actual length would be same as max length,
+            // and the following copy is avoided.
+            if (actualLength < maxLength)
+            {
+                if (actualLength == 0)
+                {
+                    tagKeys = null;
+                    tagValues = null;
+                    return;
+                }
```

Comment:
If we returned out of the block above for `if (maxLength == 0)` condition could we remove this block?